File: app/src/main/java/com/waz/zclient/pages/main/conversationpager/SecondPageFragment.java
/**
 * Wire
 * Copyright (C) 2016 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.waz.zclient.pages.main.conversationpager;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.waz.api.IConversation;
import com.waz.api.User;
import com.waz.model.ConvId;
import com.waz.model.ConversationData;
import com.waz.model.UserId;
import com.waz.zclient.OnBackPressedListener;
import com.waz.zclient.R;
import com.waz.zclient.connect.ConnectRequestFragment;
import com.waz.zclient.controllers.navigation.Page;
import com.waz.zclient.controllers.navigation.PagerControllerObserver;
import com.waz.zclient.conversation.ConversationController;
import com.waz.zclient.core.stores.connect.IConnectStore;
import com.waz.zclient.core.stores.conversation.ConversationChangeRequester;
import com.waz.zclient.pages.BaseFragment;
import com.waz.zclient.pages.main.connect.ConnectRequestLoadMode;
import com.waz.zclient.pages.main.connect.PendingConnectRequestManagerFragment;
import com.waz.zclient.pages.main.conversation.ConversationManagerFragment;
import com.waz.zclient.ui.utils.MathUtils;

import com.waz.zclient.utils.Callback;
import timber.log.Timber;

public class SecondPageFragment extends BaseFragment<SecondPageFragment.Container> implements OnBackPressedListener,
                                                                                              ConversationManagerFragment.Container,
                                                                                              PagerControllerObserver,
                                                                                              PendingConnectRequestManagerFragment.Container,
                                                                                              ConnectRequestFragment.Container {
    public static final String TAG = SecondPageFragment.class.getName();
    private static final String SECOND_PAGE_POSITION = "SECOND_PAGE_POSITION";
    public static final String ARGUMENT_CONVERSATION_ID = "ARGUMENT_CONVERSATION_ID";

    private Page currentPage;

    public static SecondPageFragment newInstance() {
        return new SecondPageFragment();
    }

    //////////////////////////////////////////////////////////////////////////////////////////
    //
    //  LifeCycle
    //
    //////////////////////////////////////////////////////////////////////////////////////////

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);
        if (isAdded()) {
            Fragment fragment = getChildFragmentManager().findFragmentById(R.id.fl__second_page_container);
            if (fragment != null) {
                fragment.setUserVisibleHint(isVisibleToUser);
            }
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        if (savedInstanceState == null) {
            currentPage = Page.NONE;
        } else {
            int pos = savedInstanceState.getInt(SECOND_PAGE_POSITION);
            currentPage = Page.values()[pos];
        }
        return inflater.inflate(R.layout.fragment_pager_second, container, false);
    }

    @Override
    public void onResume() {
        super.onResume();
        getControllerFactory().getNavigationController().addPagerControllerObserver(this);
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        outState.putInt(SECOND_PAGE_POSITION, currentPage.ordinal());
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onPause() {
        getControllerFactory().getNavigationController().removePagerControllerObserver(this);
        super.onPause();
    }

    @Override
    public void onDestroyView() {
        getControllerFactory().getNavigationController().removePagerControllerObserver(this);
        inject(ConversationController.class).removeConvChangedCallback(callback);
        super.onDestroyView();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        Fragment fragment = getChildFragmentManager().findFragmentById(R.id.fl__second_page_container);
        if (fragment != null) {
            fragment.onActivityResult(requestCode, resultCode, data);
        }
    }

    private final Callback callback = new Callback<ConversationController.ConversationChange>() {
        @Override
        public void callback(final ConversationController.ConversationChange change) {
            if (change.toConvId() == null) {
                return;
            }

            inject(ConversationController.class).withConvLoaded(change.toConvId(), new Callback<ConversationData>() {
                @Override
                public void callback(final ConversationData convData) {
                    Timber.i("Conversation: %s type: %s requester: %s",
                        change.toConvId(),
                        convData.convType(),
                        change.requester());
                    // either starting from beginning or switching fragment
                    final boolean switchingToPendingConnectRequest = (convData.convType() == IConversation.Type.WAIT_FOR_CONNECTION);
                    final boolean switchingToConnectRequestInbox = convData.convType() == IConversation.Type.INCOMING_CONNECTION;

                    if (switchingToConnectRequestInbox) {
                        Bundle arguments = new Bundle();
                        arguments.putString(ARGUMENT_CONVERSATION_ID, change.toConvId().str());
                        openPage(Page.CONNECT_REQUEST_INBOX, arguments);
                    } else if (switchingToPendingConnectRequest) {
                        Bundle arguments = new Bundle();
                        arguments.putString(ARGUMENT_CONVERSATION_ID, change.toConvId().str());
                        openPage(Page.CONNECT_REQUEST_PENDING, arguments);
                    } else {
                        openPage(Page.MESSAGE_STREAM, new Bundle());
                    }
                }
            });
        }
    };

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        inject(ConversationController.class).addConvChangedCallback(callback);
    }

    private void openPage(Page page, Bundle arguments) {
        Timber.i("openPage %s", page.name());
        if (getContainer() == null || !isResumed()) {
            return;
        }
        Fragment fragment = getChildFragmentManager().findFragmentById(R.id.fl__second_page_container);

        if (currentPage != null && currentPage.equals(page)) {
            // Scroll to a certain connect request in inbox
            if (fragment instanceof ConnectRequestFragment) {
                //TODO set a preference or something
                ((ConnectRequestFragment) fragment).setVisibleConnectRequest(new UserId(arguments.getString(ARGUMENT_CONVERSATION_ID)));
            }

            if (page != Page.CONNECT_REQUEST_PENDING) {
                return;
            }
        }
        currentPage = page;

        FragmentTransaction transaction = getChildFragmentManager().beginTransaction();
        switch (getControllerFactory().getNavigationController().getCurrentPage()) {
            case CONVERSATION_LIST:
                transaction.setCustomAnimations(R.anim.message_fade_in,
                                                R.anim.message_fade_out,
                                                R.anim.message_fade_in,
                                                R.anim.message_fade_out);
                break;
            case CONNECT_REQUEST_INBOX:
            case CONNECT_REQUEST_PENDING:
                transaction.setCustomAnimations(R.anim.fragment_animation_second_page_slide_in_from_right,
                                                R.anim.fragment_animation_second_page_slide_out_to_left);
                break;
        }

        Fragment pageFragment;
        String tag;
        switch (page) {
            case CONNECT_REQUEST_PENDING:
                getControllerFactory().getNavigationController().setRightPage(Page.PENDING_CONNECT_REQUEST_AS_CONVERSATION,
                                                                              TAG);
                pageFragment = PendingConnectRequestManagerFragment.newInstance(null,
                                                                                arguments.getString(
                                                                                    ARGUMENT_CONVERSATION_ID),
                                                                                ConnectRequestLoadMode.LOAD_BY_CONVERSATION_ID,
                                                                                IConnectStore.UserRequester.CONVERSATION);
                tag = PendingConnectRequestManagerFragment.TAG;
                break;
            case CONNECT_REQUEST_INBOX:
                getControllerFactory().getNavigationController().setRightPage(Page.CONNECT_REQUEST_INBOX, TAG);
                pageFragment = ConnectRequestFragment.newInstance(arguments.getString(ARGUMENT_CONVERSATION_ID));
                tag = ConnectRequestFragment.FragmentTag();
                break;
            case MESSAGE_STREAM:
                getControllerFactory().getNavigationController().setRightPage(Page.MESSAGE_STREAM, TAG);
                pageFragment = ConversationManagerFragment.newInstance();
                tag = ConversationManagerFragment.TAG;
                break;
            case NONE:
            default:
                return;
        }

        transaction.replace(R.id.fl__second_page_container, pageFragment, tag).commit();
    }

    @Override
    public void onOpenUrl(String url) {
        getContainer().onOpenUrl(url);
    }

    @Override
    public boolean onBackPressed() {
        Fragment fragment = getChildFragmentManager().findFragmentById(R.id.fl__second_page_container);
        return fragment instanceof OnBackPressedListener &&
            ((OnBackPressedListener) fragment).onBackPressed();
    }

    //////////////////////////////////////////////////////////////////////////////////////////
    //
    //   PagerControllerObserver
    //
    //////////////////////////////////////////////////////////////////////////////////////////

    @Override
    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
        if (position == 0 || MathUtils.floatEqual(positionOffset, 0f)) {
            getView().setAlpha(1f);
        } else {
            getView().setAlpha((float) Math.pow(positionOffset, 4));
        }
    }

    @Override
    public void onPageSelected(int position) {

    }

    @Override
    public void onPageScrollStateChanged(int state) {

    }

    //////////////////////////////////////////////////////////////////////////////////////////
    //
    //  UserProfile
    //
    //////////////////////////////////////////////////////////////////////////////////////////

    @Override
    public void onAcceptedConnectRequest(ConvId conversation) {
        Timber.i("onAcceptedConnectRequest %s", conversation);
        inject(ConversationController.class).selectConv(conversation, ConversationChangeRequester.CONVERSATION_LIST);
    }

    @Override
    public void onAcceptedPendingOutgoingConnectRequest(ConvId conversation) {
        Timber.i("onAcceptedPendingOutgoingConnectRequest %s", conversation);
        inject(ConversationController.class).selectConv(conversation, ConversationChangeRequester.CONNECT_REQUEST_ACCEPTED);
    }

    @Override
    public void dismissInboxFragment() {
        Timber.i("dismissInboxFragment");
        getControllerFactory().getNavigationController().setVisiblePage(Page.CONVERSATION_LIST, TAG);
    }

    @Override
    public void onPagerEnabledStateHasChanged(boolean enabled) {

    }

    @Override
    public void dismissUserProfile() {

    }

    @Override
    public void dismissSingleUserProfile() {

    }

    @Override
    public void showRemoveConfirmation(User user) {

    }

    public interface Container {
        void onOpenUrl(String url);
    }
}

File: app/src/main/scala/com/waz/zclient/conversationpager/SecondPageFragment.scala
/**
 * Wire
 * Copyright (C) 2018 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.waz.zclient.conversationpager

import android.content.Intent
import android.os.Bundle
import android.view.{LayoutInflater, View, ViewGroup}
import com.waz.ZLog.ImplicitTag._
import com.waz.ZLog._
import com.waz.api.IConversation
import com.waz.model.UserId
import com.waz.threading.Threading
import com.waz.zclient.common.controllers.{BrowserController, UserAccountsController}
import com.waz.zclient.connect.{ConnectRequestFragment, PendingConnectRequestManagerFragment}
import com.waz.zclient.controllers.navigation.{INavigationController, Page, PagerControllerObserver}
import com.waz.zclient.conversation.ConversationController
import com.waz.zclient.core.stores.connect.IConnectStore
import com.waz.zclient.core.stores.conversation.ConversationChangeRequester
import com.waz.zclient.pages.main.conversation.ConversationManagerFragment
import com.waz.zclient.ui.utils.MathUtils
import com.waz.zclient.{FragmentHelper, OnBackPressedListener, R}

class SecondPageFragment extends FragmentHelper
  with OnBackPressedListener
  with ConversationManagerFragment.Container
  with PagerControllerObserver
  with PendingConnectRequestManagerFragment.Container
  with ConnectRequestFragment.Container {

  import Threading.Implicits.Ui
  import SecondPageFragment._

  private lazy val navigationController   = inject[INavigationController]
  private lazy val userAccountsController = inject[UserAccountsController]
  private lazy val conversationController = inject[ConversationController]
  private lazy val browserController       = inject[BrowserController]

  override def setUserVisibleHint(isVisibleToUser: Boolean): Unit = {
    super.setUserVisibleHint(isVisibleToUser)
    if (isAdded) {
      val fragment = getChildFragmentManager.findFragmentById(R.id.fl__second_page_container)
      if (fragment != null) fragment.setUserVisibleHint(isVisibleToUser)
    }
  }

  override def onCreateView(inflater: LayoutInflater, container: ViewGroup, savedInstanceState: Bundle): View = {
    inflater.inflate(R.layout.fragment_pager_second, container, false)
  }

  override def onViewCreated(view: View, savedInstanceState: Bundle): Unit = {
    (for {
      conv <- conversationController.currentConv
      convMembers <- conversationController.currentConvMembers
    } yield conv -> convMembers.head).onUi { case (conv, userId) =>
      info(s"Conversation: ${conv.id} type: ${conv.convType}")

      // either starting from beginning or switching fragment
      val switchingToPendingConnectRequest = conv.convType == IConversation.Type.WAIT_FOR_CONNECTION
      val switchingToConnectRequestInbox = conv.convType == IConversation.Type.INCOMING_CONNECTION

      if (switchingToConnectRequestInbox) openPage(Page.CONNECT_REQUEST_INBOX, userId)
      else if (switchingToPendingConnectRequest) openPage(Page.CONNECT_REQUEST_PENDING, userId)
      else openPage(Page.MESSAGE_STREAM, userId)
    }

  }

  override def onActivityResult(requestCode: Int, resultCode: Int, data: Intent): Unit = {
    super.onActivityResult(requestCode, resultCode, data)
    val fragment = getChildFragmentManager.findFragmentById(R.id.fl__second_page_container)
    if (fragment != null) fragment.onActivityResult(requestCode, resultCode, data)
  }

  private def openPage(page: Page, userId: UserId) = {
    info(s"openPage ${page.name} userId $userId")

    Some(page)
      .collect {
        case Page.CONNECT_REQUEST_PENDING =>
          import PendingConnectRequestManagerFragment._
          navigationController.setRightPage(Page.PENDING_CONNECT_REQUEST_AS_CONVERSATION, Tag)
          Tag -> newInstance(userId, IConnectStore.UserRequester.CONVERSATION)
        case Page.CONNECT_REQUEST_INBOX =>
          import ConnectRequestFragment._
          navigationController.setRightPage(Page.CONNECT_REQUEST_INBOX, SecondPageFragment.Tag)
          FragmentTag -> newInstance(userId)
        case Page.MESSAGE_STREAM =>
          import ConversationManagerFragment._
          navigationController.setRightPage(Page.MESSAGE_STREAM, SecondPageFragment.Tag)
          Tag -> newInstance
      }
      .map { case (tag, pageFragment) =>
        getChildFragmentManager.beginTransaction.replace(R.id.fl__second_page_container, pageFragment, tag)
      }
      .map(_ -> navigationController.getCurrentPage)
      .collect {
        case (transaction, Page.CONVERSATION_LIST) =>
          transaction.setCustomAnimations(
            R.anim.message_fade_in,
            R.anim.message_fade_out,
            R.anim.message_fade_in,
            R.anim.message_fade_out
          )
        case (transaction, Page.CONNECT_REQUEST_INBOX | Page.CONNECT_REQUEST_PENDING) =>
          transaction.setCustomAnimations(
            R.anim.fragment_animation_second_page_slide_in_from_right,
            R.anim.fragment_animation_second_page_slide_out_to_left
          )
      }
      .foreach(transaction => transaction.commit())
  }

  override def onOpenUrl(url: String): Unit = browserController.openUrl(url)

  override def onBackPressed: Boolean = {
    val fragment = getChildFragmentManager.findFragmentById(R.id.fl__second_page_container)
    fragment.isInstanceOf[OnBackPressedListener] && fragment.asInstanceOf[OnBackPressedListener].onBackPressed
  }

  override def onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int): Unit = {
    if (position == 0 || MathUtils.floatEqual(positionOffset, 0f)) getView.setAlpha(1f)
    else getView.setAlpha(Math.pow(positionOffset, 4).toFloat)
  }

  override def onPageSelected(position: Int): Unit = {
  }

  override def onPageScrollStateChanged(state: Int): Unit = {
  }

  override def onAcceptedConnectRequest(userId: UserId): Unit = {
    info(s"onAcceptedConnectRequest $userId")
    userAccountsController.getConversationId(userId).flatMap { convId =>
      conversationController.selectConv(convId, ConversationChangeRequester.CONVERSATION_LIST)
    }
  }

  override def dismissInboxFragment(): Unit = {
    info("dismissInboxFragment")
    navigationController.setVisiblePage(Page.CONVERSATION_LIST, Tag)
  }

  override def onPagerEnabledStateHasChanged(enabled: Boolean): Unit = {
  }

  override def dismissUserProfile(): Unit = {
  }

  override def dismissSingleUserProfile(): Unit = {
  }

  override def showRemoveConfirmation(userId: UserId): Unit = {
  }
}

object SecondPageFragment {
  val Tag: String = classOf[SecondPageFragment].getName
  val ArgConversationId = "ARGUMENT_CONVERSATION_ID"

  def newInstance = new SecondPageFragment
}
